mipengine/controller/algorithm_executor.py

```diff
@@ -108,27 +113,81 @@ def __init__(
         )
 
     def run(self):
-        algorithm_result = self.algorithm_flow_module.run(self.execution_interface)
+        try:
+            algorithm_result = self.algorithm_flow_module.run(self.execution_interface)
+            return algorithm_result
 
-        self.clean_up()
+        except SoftTimeLimitExceeded as stle:
+            error_result = ErrorResult(
+                error_title="SoftTimeLimitExceeded", description=str(stle.args[0])
+            )
+            # TODO logging
+            print(f"{error_result=}")
+            return error_result
+        except TimeLimitExceeded as tle:
+            print(f"{tle=}")
+            error_result = ErrorResult(
+                error_title="TimeLimitExceeded", description=str(tle.args[0])
+            )
+            # TODO logging
+            print(f"{error_result=}")
+            return error_result
+        except TimeoutError as te:
+            print(f"{te=}")
+            error_result = ErrorResult(
+                error_title="TimeoutError", description=str(te.args[0])
+            )
+            # TODO logging
+            print(f"{error_result=}")
+            return error_result
+        except:
+            import traceback
 
-        return algorithm_result
+            print(traceback.format_exc())
+        finally:
+            self.clean_up()
```

Comment:
If a call fails with timeout and theNode service is down, the clean_up will most probably fail as well.

Reply:
Yes, the node that is down cannot execute anything, the cleanup is for the rest of the nodes that are still up and listening.

Reply:
But the cleanup will fail again and raise an exception that is not handled. So the rest of the nodes will not be cleaned up and the exception will not be handled.

Reply:
Where do you base that scenario? These exceptions are raised when the `get(..)` method is called. Any task that does not need a result does not call the `get()` method, only the `delay(..)`, `clean_up` is this kind of task.

The `delay(..)` method:
- non-blocking, returns immediately
- queues the task and the worker on the other side starts executing it
- no timeout related exceptions are raised
- returns an AsyncResult object

The `get(..)` method:
- blocking, will return only when the result of the task is on the backend
- can potentially raise one of the 3  aforementioned timeout related exceptions

Reply:
Sorry!!! Yes, you are right, I thought when cleaning up, we were waiting for it to finish.

Reply:
What if the node is down completely though, rabbitmq as well, and the delay() call fails?

Reply:
sidenote: I have erroneously decorated the `clean_up` and `queue_run_udf` with `@time_limit_exceeded_handler`, I ll remove it.

Concerning rabbitmq being down while `delay` is called will need a whole different handling which will not be addressed in this pr, it is my mistake that I have not made that clear from the beginning but I was not sure about this when I started implementing this solution. I added a comment about that in the current ticket description and at some point I ll create a new ticket about that.

Reply:
Ok! Please add the ticket to the epic so that it's clear that the story is blocked by it.

sidenote: It would really be helpful to add e2e_tests handling cases where a node is down. This is a very usual scenario in MIP and it's easy to do that with the automated tests we have. Would you consider adding a ticket for that as well?

Reply:
[MIP-424](https://team-1617704806227.atlassian.net/browse/MIP-424)
[MIP-425](https://team-1617704806227.atlassian.net/browse/MIP-425)
I ll add them to the epics as well..